### ISSUE TITLE
Terminal reset needed after calling Delta with a wrong number of file arguments

### ISSUE THREAD

**Issue by lepotic** [dandavison/delta#513]
Hi, thanks for making Delta!

I just installed it for the first time and tried to see if it would beautify an already stored diff in a .patch file.
After printing the usage error message, it left the terminal in a state with local echo disabled.
(The same happened when calling it without any arguments too.)

Here's a before/after stty(1) output:

```
; echo $TERM
st-256color
; stty
speed 38400 baud; line = 0;
-brkint -imaxbel
; delta
Usage: delta minus_file plus_file
speed 38400 baud; line = 0;
lnext = <undef>; min = 1; time = 0;
-brkint -imaxbel
tab3
-icanon -echo -echoe -echok
;
```

```
; echo $TERM
xterm
; stty
speed 38400 baud; line = 0;
erase = ^H;
-brkint -imaxbel iutf8
; delta
Usage: delta minus_file plus_file
speed 38400 baud; line = 0;
erase = ^H; lnext = <undef>; min = 1; time = 0;
-brkint -imaxbel iutf8
tab3
-icanon -echo -echoe -echok
;
```

Versions:

```
; uname -a
Linux archlinux 5.10.11-arch1-1 #1 SMP PREEMPT Wed, 27 Jan 2021 13:53:16 +0000 x86_64 GNU/Linux
; pacman -Q git-delta xterm st bash
git-delta 0.5.1-2
xterm 363-1
st 0.8.4
bash 5.1.004-1
```

It was built from this AUR package - https://aur.archlinux.org/packages/git-delta

Initially the issue happened 100% of the time in all the terminals I tried, but later it turned out that sometimes the terminal was fine. Sometimes it even "alternated" in the same terminal instance.

**Comment by Kr1ss-XD** [comment]
FWIW, I was able to replicate the terminal crash after running `delta` without arguments or with only one filename argument.

Both in Tmux and without this is, and both with xterm and xfce4-terminal.

`delta <file1> <file2> <file3> [...]` fails gracefully with an error message and leaves the terminal unaffected.

**Comment by foundObjects** [comment]
Also affected by this, running `tput init` in the broken terminal fixes things.

**Comment by dandavison** [comment]
Hi all, I haven't been able to reproduce this yet, but I've just been trying things on MacOS (iTerm2, Terminal.app, xterm, and VSCode xterm.js) whereas I think you're all on Linux. I should dig out my Raspberry Pi.

I've just merged https://github.com/dandavison/delta/pull/546 which switches to call `git diff file_A file_B`  instead of `diff -u file_A file_B`, and generally affects the relevant code, so it would be interesting to know if it still occurs on master.

**Comment by lepotic** [comment]
Still happens, even after 53e59f2. It's no big deal, though, really.
I only stumbled upon this because it was my first time trying to use Delta and, thinking that it is a _pure_ filter sitting between some diff output and less(1), I expected it to work on stdin (or explicit `-` filename) or a file.

**Comment by dandavison** [comment]
OK, thanks a lot for testing.

> thinking that it is a pure filter sitting between some diff output and less(1), I expected it to work on stdin (or explicit - filename) or a file.

Right, we could implement the singe-file-argument variant -- as you say, that would be consistent with general unix CLI style.

**Comment by vrza** [comment]
**When stdout is a tty, running the program without any arguments results in changed tty line settings, including disabled echo of input characters.**

Confirmed with GNU Bash 5.1 on gnome-terminal, xterm, st, kitty and virtual terminals of the Linux console. Most likely every VT100 compatible terminal and terminal emulator is affected, but some shells might work around the issue, making it less apparent or masking it completely (see details below).

Steps to reproduce, using Bash 5.1:

```
$ stty -a > before
$ delta
$ stty -a > after
```

Notice that keyboard input to the terminal is not echoed after `delta` exits. Compare `stty(1)` outputs before and after. `tput init` or `reset` is needed to restore the terminal to the initial state.

**If stdout is not a tty (but stderr and stdin might be) the bug is not triggered!** This is interesting, because in the problematic scenario of no arguments, the program outputs usage to stderr, and nothing to stdout!. E.g. running:

```
$ delta > /dev/null
```

will not result in changed tty settings.

Behavior of tested shells other than bash 5.1:

- tcsh 6.22 seems to restore all the tty settings back to the default state
- zsh 5.8 seems to restore only the echo setting

Another, possibly unrelated, thing I noticed while looking into this -- the program always seems to run `less` in the background, presumably to detect the installed version. This seems like an unnecessary step in the case when the only output are help instructions.

Complete list of changed line setings (as reported by `stty(1)`):

```
lnext = ^               ->  lnext = <undef>
tab0                    ->  tab3
icanon echo echoe echok ->  -icanon -echo -echoe -echok 
```

Their meanings are (from the `stty(1)` man page):

```
lnext CHAR        CHAR will enter the next character quoted
[-]icanon:        enable/disable special characters: erase, kill, werase, rprnt
tabN:             horizontal tab delay style, N in [0..3]
[-]echo           echo input characters
[-]echoe          same as [-]crterase
[-]crterase       echo erase characters as backspace-space-backspace
[-]echok          echo a newline after a kill character
```
